TMS-231: Use (mostly) only one query to populate all areas within a center page.

witsch opened the pull request:
See [TMS-231](https://zeit-online.atlassian.net/browse/TMS-231).

witsch committed:
TMS-231: Factor out actual TMS request and its error handling.

witsch committed:
TMS-231: Use only one (cached) query to populate all areas within the (center)page.

Note that the query currently fetches 20 items so areas will remain empty if
the page requires more than that.  This will be addressed separately...

witsch committed:
TMS-231: Filter duplicates inline so that we always get the desired number if items.

witsch committed:
TMS-231: Clean up the code a bit.

witsch committed:
TMS-231: TMS-231: TMS-231: Fetch more items if the current result set has been exhausted.

witsch committed:
TMS-231: TMS-231: Stop if the TMS has no more hits.

witsch committed:
TMS-231: Estimate the number of results upfront so only one query is needed.

We add a few results in case there are manually 'pinned' teasers on the page.

witsch committed:
TMS-231: TMS-231: Simplify and prepare fixing `total_hits`.

witsch committed:
TMS-231: Fix `total_hits` used for "ranking" area.

The "ranking" area (instance) is a separate object, which only
worked, because previously a search was performed unconditionally
for each called area.  Now we use the cached value for 'hits'.

witsch committed:
TMS-231: Adjust test to changed call from bd6e53fa.

The `id`/`topicpage` is now passed as a keyword.

wosc commented on src/zeit/content/cp/automatic.py:
```diff
@@ -294,35 +294,65 @@ class TMSContentQuery(ContentQuery):
         result = []
-        topicpage = self.context.referenced_topicpage
+        cp = zeit.content.cp.interfaces.ICenterPage(self.context)
+        rows = cp.total_teaser_count() + 5      # total teasers + some spares
```
Das scheint mir als Heuristik etwas zu grob. Wenn wir uns schon zum Sonderfall "es geht um die gleiche TMS-Topicpage" herablassen, sollten wir denke ich dann auch nur die Areas berücksichtigen, die automatic_type=topicpage und referenced_topicpage=self.referenced_topicpage haben. Insofern würd ich eher `cp._cached_areas` irgendwie public machen und dann hier benutzen, anstatt auf der CP `total_teaser_count` einzuführen.

Und... da du unten eh hast "hol so lang welche nach bis es genug sind", kann man die magischen +5 doch auch sein lassen, oder was für einen 80%-Fall stellst du dir da vor?

witsch replied:
Ja, da hast du Recht, daß es natürlich auch noch andere areas gibt und wir nur die zählen könnten, die hier relevant sind.  Da hatte ich tatsächlich (nur) das Campus Beispiel im Kopf, wo's halt nur diese gab, also mehr oder weniger... 🙂  Und dann hat der Helper auf der centerpage nichts zu suchen.  Das ist jedenfalls in c3deecee2470440989c4a9143ef58b8236d37be7 & 16fa52cecf85cc5b9bf270e0958ff9c2060de6ed gefixt...

Und ja, der "80%-Fall", den ich mir tatsächlich vorgestellt hab, ist das explizite "anpinnen" von ein paar (wenigen) Artikeln.  Wobei die ja dann in der aktuellen Version eigentlich auch schon mitgezählt würden, oder?

wosc commented on src/zeit/content/cp/automatic.py:
```diff
@@ -294,35 +294,65 @@ class TMSContentQuery(ContentQuery):
+        cp = zeit.content.cp.interfaces.ICenterPage(self.context)
+        rows = cp.total_teaser_count() + 5      # total teasers + some spares
+        cache = cp._topic_queries
+        key = (self.topicpage, self.filter_id, self.start)
```
Noice. :)

wosc commented on src/zeit/content/cp/automatic.py:
```diff
@@ -294,35 +294,65 @@ class TMSContentQuery(ContentQuery):
+        if key in cache:
+            _, _, hits = cache[key]
+        else:
+            _, hits = iself._get_documents(start=self.start, rows=0)
```
`iself`? Du vim-User, du.

witsch replied:
Hihi, offensichtlich wurde `total_hits` (wie vermutet) immer erst nach den Areas aufgerufen :smile: — bzw. gab es überall mindestens eine.  Ist aber mit 6c490e26eb1c4a2250123b9c816beb2af8bfc568 gefixt & getestet...

wosc commented on src/zeit/content/cp/automatic.py:
```diff
@@ -294,35 +294,65 @@ class TMSContentQuery(ContentQuery):
+            response, start, _ = cache[key]
+        else:
+            start = self.start
+            response, hits = self._get_documents(start=start, rows=rows)
```
Statt `(response, hits)` gibts `zeit.cms.interfaces.Result`, in der Hoffnung dass es ein bisschen weniger tupelig zu lesen+schreiben ist.

witsch replied:
Naja, das Problem hier ist ja, daß ich 'nen Iterator haben will, und — hatte ich vorher allerdings auch noch nicht probiert — kann ich keine extra Attribute setzen:
```
~/Z/f/w/s/zeit.content.cp (TMS-231|✚1) $ python
Python 2.7.15 (default, May  1 2018, 16:44:08) 
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> i = iter([])
>>> i.hits = 9
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'listiterator' object has only read-only attributes (assign to .hits)
```
:cry:

wosc commented on src/zeit/content/cp/automatic.py:
```diff
@@ -294,35 +294,65 @@ class TMSContentQuery(ContentQuery):
+            try:
+                item = response.next()
+            except StopIteration:
+                start = start + rows            # fetch next batch
```
Ahaaa. _Das_ ist der eigentliche Fix für das TMS-Dublettenproblem, nicht wahr. Dass man die TMS-Results außerdem noch cached ist mehr so die (campus-)cherry on top, oder?

Und jetzt wo ich das so sehe, erinner ich mich auch: [wir hatten sowas schonmal](https://github.com/ZeitOnline/zeit.content.cp/commit/2c7e5f2ce327a3910cd480d6c79212b53f93b9b6), wenn auch nicht so schön klar aufgeschrieben wie du jetzt.

witsch replied:
Genau, der Iterator eben.  Und wirklich gecached sind die ja eigentlich nicht, nur eben die (iterierbaren) Resultatlisten zentral gespeichert.  Und dann noch vorher einmal halbwegs "schlau" raten (also zählen :innocent:), um nicht unnötig oft Nachschlag holen zu müssen...

witsch replied:
Also okay, die Resultate quasi nicht, die Resultatquellen natürlich schon...  immer diese pedantischen Neukollegen, ne!? :joy:

witsch committed:
TMS-231: Refactor area cache to be "publically" available.

This way TMS queries can count only in relevant areas; refs 97045160.

witsch committed:
TMS-231: Fix 97045160 and only count teasers in relevant areas.

witsch committed:
TMS-231: Fix 7193cf87 and make sure `total_hits` can be called first.

witsch committed:
TMS-231: Cache area `values()` so they don't consume extra TMS results.

`filter_values` for example iterates the area's `values()` again when
called via `existing_teasers`. This has simply cause additional queries
before, but with the centrally cached result sets, this now consumes
results, effectively changing the order in which they are assigned to
the page's (automatic) areas.

Please note that this also introduces a central cache on the centerpage,
which can then be used to cache the various things we need (on the
transaction boundary).

witsch committed:
TMS-231: Adjust tests for automatic areas after a39c18e4.

These tests alter areas after `IRenderedArea.values()` has been called.
Since its results are now cached and the tested changes only need to
be visible with the next request/transaction, a commit can be used to
safely invalidate the cache.

witsch committed:
TMS-231: Add (zope.)security declaration for the cache from a39c18e4.

witsch committed:
TMS-231: Add decorator to cache data on a centerpage.

The results of the decorated function will be stored in a dictionary on
the centerpage instance and invalidated at the transaction boundary. The
dictionary keys are built using the optional `keyfunc`, which is called
with `self` as a single argument. By default the instances `__name__`
attribute is used, which works nicely for our (automatic) areas.

witsch committed:
TMS-231: Allow to specify the cache attribute to be able to avoid conflicts.

witsch committed:
TMS-231: Simplify cache setup for topic queries.

witsch committed:
TMS-231: Refactor `teasered_content_above` & `manual_content_below` helpers.

They are only used by `existing_teasers` and can therefore be "inlined".
Note that the contents of the "above" and manual teasers are still
cached so they are not need rendered multiple times.

witsch committed:
TMS-231: Simplify `existing_teasers` after 87417b6b.

witsch committed:
TMS-231: Further simplify and use common cache for content areas as well.

witsch commented:
Hmm, also ich glaube ich könnte noch ne Weile weiter putzen, aber nachdem der Kernpunkt ja erledigt und auch der Code imho etwas lesbarer ist, hör ich jetzt mal auf.  Die Cache-Strategien können wir ja ein andermal weiter verfolgen... 🙂

wosc committed:
pep8

wosc committed:
Invert double negative

wosc committed:
Clarify naming

wosc committed:
Clarify naming

wosc committed:
TMS-231: Be even more precise which areas we count as belonging to the same topicpage

wosc reviewed (approved):
![excellent](https://media.giphy.com/media/uL2GwdqRPoqnS/giphy.gif)

Ich hab an ein paar Namen noch etwas rumgeschraubt (vmtl brauchte ich das für mein persönliches Verständnis...), ich hoffe du bist mir nicht böse.

wosc committed:
Merge branch 'master' into TMS-231

wosc committed:
Update changelog
